api/src/test/java/org/openmrs/module/sync/SyncOrderSetAttributeTest.java

```diff
@@ -0,0 +1,108 @@
+                ORIGINAL_COUNT = orderSetService.getAllOrderSetAttributeTypes().size();
+                OrderSetAttributeType orderSetAttributeType = new OrderSetAttributeType();
+                orderSetAttributeType.setName("Surgery");
+                orderSetAttributeType.setDatatypeClassname(FreeTextDatatype.class.getName());
+                orderSetService.saveOrderSetAttributeType(orderSetAttributeType);
+                Assertions.assertNotNull(orderSetAttributeType.getId());
+                Assertions.assertEquals(ORIGINAL_COUNT + 1, orderSetService.getAllOrderSetAttributeTypes().size() );
+            }
+
+            @Override
+            public void changedBeingApplied(List<SyncRecord> syncRecords, Record record) throws Exception {
+                super.changedBeingApplied(syncRecords, record);
+            }
+
+            public void runOnParent() throws Exception {
+                OrderSetService orderSetService = Context.getOrderSetService();
+                Assertions.assertEquals(ORIGINAL_COUNT + 1, orderSetService.getAllOrderSetAttributeTypes().size() );
+            }
+        });
+    }
+
+    @Test
+    @Transactional(propagation = Propagation.NOT_SUPPORTED)
+    public void shouldSyncOrderSetAttribute() throws Exception {
+        runSyncTest(new SyncTestHelper() {
+            String NEW_ORDER_SET_ATTRIBUTE_UUID = null;
+            public void runOnChild() throws Exception {
+
+                OrderSetService orderSetService = Context.getOrderSetService();
+                OrderSetAttributeType attributeType = orderSetService.getOrderSetAttributeTypeByName("Special order");
+                Assertions.assertNotNull(attributeType);
+
+                OrderSet orderSet = new OrderSet();
+                orderSet.setOperator(OrderSet.Operator.ALL);
+                orderSet.setName("NewOrderSet");
+                orderSet.setDescription("New Order Set");
+                OrderSetMember orderSetMember = new OrderSetMember();
+                orderSetMember.setOrderType(Context.getOrderService().getOrderType(1));
+                orderSetMember.setConcept(Context.getConceptService().getConcept(3));
+                orderSetMember.setCreator(attributeType.getCreator());
+                orderSetMember.setDateCreated(new Date());
+                orderSetMember.setRetired(false);
+                List<OrderSetMember> orderSetMembers = new ArrayList<>(Arrays.asList(orderSetMember));
+                orderSet.setOrderSetMembers(orderSetMembers);
+                OrderSetAttribute orderSetAttribute = new OrderSetAttribute();
+                orderSetAttribute.setAttributeType(attributeType);
+                orderSetAttribute.setOrderSet(orderSet);
+                orderSetAttribute.setValueReferenceInternal("order set 1");
+
+                orderSet.addAttribute(orderSetAttribute);
+                orderSet = orderSetService.saveOrderSet(orderSet);
```

Comment:
@mseaton , I think I ran into a real problem with sync. The OrderSet is saved with no errors, and as far as I can tell the object looks OK, but the sync throws the following exception:
`
DEBUG - SyncUtil.setProperty(226) |2023-03-27T16:02:29,463| Successfully called setUuid(eb37064b-3f37-4b3c-ace9-38d5048201bc)
DEBUG - SyncIngestServiceImpl.processOpenmrsObject(548) |2023-03-27T16:02:29,464| Trying to set property:
DEBUG - SyncIngestServiceImpl.processOpenmrsObject(549) |2023-03-27T16:02:29,464| Field = null
DEBUG - SyncIngestServiceImpl.processOpenmrsObject(550) |2023-03-27T16:02:29,464| xmlFieldName = _org.openmrs.OrderSet.orderSetMembersIndexBackref
DEBUG - SyncIngestServiceImpl.processOpenmrsObject(551) |2023-03-27T16:02:29,464| xmlFieldValue = 0
DEBUG - SyncIngestServiceImpl.processOpenmrsObject(552) |2023-03-27T16:02:29,464| xmlFieldType = integer
DEBUG - SyncUtil.valForField(373) |2023-03-27T16:02:29,464| field is null, trying to get class from xmlFieldType: integer
DEBUG - SyncUtil.valForField(377) |2023-03-27T16:02:29,465| attempting to determine type based on xmlFieldType: integer
ERROR - SyncIngestServiceImpl.processOpenmrsObject(562) |2023-03-27T16:02:29,465| Error when trying to set _org.openmrs.OrderSet.orderSetMembersIndexBackref in className org.openmrs.OrderSetMember
java.lang.IllegalArgumentException: Unable to load class: integer
	at org.openmrs.module.sync.SyncUtil.valForField(SyncUtil.java:382) ~[classes/:?]
	at org.openmrs.module.sync.SyncUtil.setProperty(SyncUtil.java:221) ~[classes/:?]
	at org.openmrs.module.sync.api.impl.SyncIngestServiceImpl.processOpenmrsObject(SyncIngestServiceImpl.java:558) ~[classes/:?]
	at org.openmrs.module.sync.api.impl.SyncIngestServiceImpl.processSyncItem(SyncIngestServiceImpl.java:431) ~[classes/:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_151]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_151]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_151]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_151]
`
Do you think this is a problem with sync or with the way the OrderSet object is constructed? Thanks!

Reply:
Hi @cioan .  So I looked into this, and the issue is that Sync does not support properties of OpenMRS Objects that are Lists, only Sets.  This must be the only place that List is used.  I'd punt on this for now.  Leave the test, but add an Ignore annotation so that it doesn't run, and add a comment explaining that it is ignore because OrderSets are not supported by sync as sync is not yet able to handle lists, and if / when this is fixed in the future, this test should be unignored.